File: mobile/lib/models/mandate.dart
/// User Mandate — client model (just the editable surface for Settings).
/// Mirrors backend/app/schemas/mandate.py.
library;

class ComplianceFlags {
  const ComplianceFlags({
    this.halal = false,
    this.esgLite = false,
    this.noTobaccoAlcoholGambling = false,
    this.noFossilFuels = false,
    this.longOnly = true,
    this.liquidOnly = true,
    this.tickerBlocklist = const [],
    this.tickerAllowlist,
    this.customConstraints = const [],
  });

  final bool halal;
  final bool esgLite;
  final bool noTobaccoAlcoholGambling;
  final bool noFossilFuels;
  final bool longOnly;
  final bool liquidOnly;
  final List<String> tickerBlocklist;
  final List<String>? tickerAllowlist;
  final List<String> customConstraints;

  ComplianceFlags copyWith({
    bool? halal,
    bool? esgLite,
    bool? noTobaccoAlcoholGambling,
    bool? noFossilFuels,
    bool? longOnly,
    bool? liquidOnly,
    List<String>? tickerBlocklist,
    List<String>? tickerAllowlist,
    bool clearAllowlist = false,
  }) {
    return ComplianceFlags(
      halal: halal ?? this.halal,
      esgLite: esgLite ?? this.esgLite,
      noTobaccoAlcoholGambling:
          noTobaccoAlcoholGambling ?? this.noTobaccoAlcoholGambling,
      noFossilFuels: noFossilFuels ?? this.noFossilFuels,
      longOnly: longOnly ?? this.longOnly,
      liquidOnly: liquidOnly ?? this.liquidOnly,
      tickerBlocklist: tickerBlocklist ?? this.tickerBlocklist,
      tickerAllowlist: clearAllowlist ? null : (tickerAllowlist ?? this.tickerAllowlist),
      customConstraints: customConstraints,
    );
  }

  Map<String, dynamic> toPatchJson() => {
        'halal': halal,
        'esg_lite': esgLite,
        'no_tobacco_alcohol_gambling': noTobaccoAlcoholGambling,
        'no_fossil_fuels': noFossilFuels,
        'long_only': longOnly,
        'liquid_only': liquidOnly,
        'ticker_blocklist': tickerBlocklist,
        if (tickerAllowlist != null) 'ticker_allowlist': tickerAllowlist,
      };

  factory ComplianceFlags.fromJson(Map<String, dynamic> j) {
    return ComplianceFlags(
      halal: (j['halal'] as bool?) ?? false,
      esgLite: (j['esg_lite'] as bool?) ?? false,
      noTobaccoAlcoholGambling: (j['no_tobacco_alcohol_gambling'] as bool?) ?? false,
      noFossilFuels: (j['no_fossil_fuels'] as bool?) ?? false,
      longOnly: (j['long_only'] as bool?) ?? true,
      liquidOnly: (j['liquid_only'] as bool?) ?? true,
      tickerBlocklist: ((j['ticker_blocklist'] as List?) ?? const []).cast<String>(),
      tickerAllowlist: (j['ticker_allowlist'] as List?)?.cast<String>(),
    );
  }
}

/// Specific dollar/year target. Optional — only set for the "specific_goal" path.
class TargetOutcome {
  const TargetOutcome({
    required this.amount,
    this.currency = 'USD',
    required this.byYear,
  });

  final double amount;
  final String currency;
  final int byYear;

  factory TargetOutcome.fromJson(Map<String, dynamic> j) => TargetOutcome(
        amount: (j['amount'] as num).toDouble(),
        currency: j['currency'] as String? ?? 'USD',
        byYear: (j['by_year'] as num).toInt(),
      );

  Map<String, dynamic> toJson() => {
        'amount': amount,
        'currency': currency,
        'by_year': byYear,
      };
}

/// Granular risk profile (1-5 / -1..1 scales). Derived from the risk-scenario
/// questions during onboarding. Used by safety floor + drift detection.
class RiskComponents {
  const RiskComponents({
    required this.drawdownResponse,
    required this.regretAsymmetry,
    required this.concentrationTolerance,
  });

  /// 1 (panic) → 5 (hold steady).
  final int drawdownResponse;

  /// -1 (regret losses more) → 0 (neutral) → 1 (regret missed gains more).
  final int regretAsymmetry;

  /// 1 (diversified always) → 5 (high-conviction concentration ok).
  final int concentrationTolerance;

  factory RiskComponents.fromJson(Map<String, dynamic> j) => RiskComponents(
        drawdownResponse: (j['drawdown_response'] as num?)?.toInt() ?? 3,
        regretAsymmetry: (j['regret_asymmetry'] as num?)?.toInt() ?? 0,
        concentrationTolerance:
            (j['concentration_tolerance'] as num?)?.toInt() ?? 3,
      );

  Map<String, dynamic> toJson() => {
        'drawdown_response': drawdownResponse,
        'regret_asymmetry': regretAsymmetry,
        'concentration_tolerance': concentrationTolerance,
      };
}

/// BL12 holdings-vs-mandate audit result. CR101-MOBILE calls this
/// immediately after a risk-limit PATCH to render the retro-tightening
/// disclosure — flag, never a forced sell (`GET /v1/mandate/{id}/audit`).
class HoldingsAuditResult {
  const HoldingsAuditResult({
    required this.passed,
    required this.mandateVersion,
    required this.maxOpenPositionsBreach,
    required this.maxOpenRiskPctBreach,
    required this.violationTickers,
  });

  final bool passed;
  final int mandateVersion;
  final bool maxOpenPositionsBreach;
  final bool maxOpenRiskPctBreach;
  final List<String> violationTickers;

  bool get anyRetroBreach => maxOpenPositionsBreach || maxOpenRiskPctBreach;

  factory HoldingsAuditResult.fromJson(Map<String, dynamic> j) {
    final violations = (j['violations'] as List? ?? const [])
        .cast<Map>()
        .map((v) => v['ticker'] as String? ?? '')
        .where((t) => t.isNotEmpty)
        .toList();
    return HoldingsAuditResult(
      passed: (j['passed'] as bool?) ?? true,
      mandateVersion: (j['mandate_version'] as num?)?.toInt() ?? 0,
      maxOpenPositionsBreach: (j['max_open_positions_breach'] as bool?) ?? false,
      maxOpenRiskPctBreach: (j['max_open_risk_pct_breach'] as bool?) ?? false,
      violationTickers: violations,
    );
  }
}

/// CR129 — the enforced value for each of the seven risk limits, resolved
/// server-side: the explicit override when one is set, else the risk-profile
/// preset. Stamped onto GET/PATCH /v1/mandate responses by the backend
/// (DEF193 + d1d4076d); never persisted, never computed client-side (CR046).
/// Every field is nullable because an installed build may talk to an older
/// backend: pre-DEF193 sends no `resolved` at all, and alpha-2026-08-19-2
/// carries only the two CR101-BE1 caps. A null field means UNKNOWN — the UI
/// falls back to no-number copy, it never fabricates a preset.
class ResolvedCaps {
  const ResolvedCaps({
    this.sectorCapPct,
    this.singleNameCapPct,
    this.maxOpenPositions,
    this.postLossCooldownHours,
    this.maxTradesPerDay,
    this.maxTradesPerWeek,
    this.maxOpenRiskPct,
  });

  final double? sectorCapPct;
  final double? singleNameCapPct;
  final int? maxOpenPositions;
  final double? postLossCooldownHours;
  final int? maxTradesPerDay;
  final int? maxTradesPerWeek;
  final double? maxOpenRiskPct;

  factory ResolvedCaps.fromJson(Map<String, dynamic> j) {
    return ResolvedCaps(
      sectorCapPct: (j['sector_cap_pct'] as num?)?.toDouble(),
      singleNameCapPct: (j['single_name_cap_pct'] as num?)?.toDouble(),
      maxOpenPositions: (j['max_open_positions'] as num?)?.toInt(),
      postLossCooldownHours:
          (j['post_loss_cooldown_hours'] as num?)?.toDouble(),
      maxTradesPerDay: (j['max_trades_per_day'] as num?)?.toInt(),
      maxTradesPerWeek: (j['max_trades_per_week'] as num?)?.toInt(),
      maxOpenRiskPct: (j['max_open_risk_pct'] as num?)?.toDouble(),
    );
  }

  /// The resolved value keyed by the backend PATCH field name — the Settings
  /// render path is key-driven (see `kRiskLimitFields`).
  num? valueFor(String key) {
    switch (key) {
      case 'sector_cap_pct':
        return sectorCapPct;
      case 'single_name_cap_pct':
        return singleNameCapPct;
      case 'max_open_positions':
        return maxOpenPositions;
      case 'post_loss_cooldown_hours':
        return postLossCooldownHours;
      case 'max_trades_per_day':
        return maxTradesPerDay;
      case 'max_trades_per_week':
        return maxTradesPerWeek;
      case 'max_open_risk_pct':
        return maxOpenRiskPct;
      default:
        return null;
    }
  }
}

/// CR129 items 2-3 — the Day Trader preset as the SERVER defines it, stamped
/// onto GET/PATCH /v1/mandate responses (568bd360). `overrides` is the exact
/// PATCH body that applies the preset — key = mandate field name, value = the
/// permissive setting. The picker PATCHes it back VERBATIM: the backend's
/// `is_day_trader_preset()` recognises that payload and journals the CR131
/// cohort marker, so the raw decoded nums are kept untouched — never rounded,
/// never re-typed, and never duplicated as client constants (the CR129-MOBILE
/// fence: "Do NOT hard-code any preset value or cap — server-sourced only").
/// `disclosure` is the selection-time dialog body, rendered verbatim.
class DayTraderPresetInfo {
  const DayTraderPresetInfo({required this.overrides, required this.disclosure});

  final Map<String, num> overrides;
  final String disclosure;

  factory DayTraderPresetInfo.fromJson(Map<String, dynamic> j) {
    return DayTraderPresetInfo(
      overrides: ((j['overrides'] as Map?) ?? const {})
          .map((k, v) => MapEntry(k as String, v as num)),
      disclosure: j['disclosure'] as String? ?? '',
    );
  }
}

class UserMandate {
  const UserMandate({
    required this.userId,
    required this.version,
    required this.displayName,
    required this.locale,
    required this.timezone,
    required this.primaryGoal,
    required this.horizon,
    this.targetOutcome,
    required this.path,
    required this.riskScore,
    required this.riskComponents,
    this.riskQuotes = const [],
    required this.maxDrawdownPct,
    required this.learningStyle,
    required this.compliance,
    required this.plan,
    this.trialExpiresAt,
    required this.creditBalance,
    this.creditAllowance = 0,
    this.roomCost = 0,
    this.creditsResetAt,
    this.roomCooldownUntil,
    this.createdAt,
    this.updatedAt,
    this.sectorCapPct,
    this.singleNameCapPct,
    this.postLossCooldownHours,
    this.maxOpenPositions,
    this.maxTradesPerDay,
    this.maxTradesPerWeek,
    this.maxOpenRiskPct,
    this.resolved,
    this.dayTraderPreset,
  });

  final String userId;
  final int version;
  final String displayName;
  final String locale;
  final String timezone;
  final String primaryGoal;
  final String horizon;
  final TargetOutcome? targetOutcome;
  final String path;
  final int riskScore;
  final RiskComponents riskComponents;
  final List<String> riskQuotes;
  final int maxDrawdownPct;
  // The seven settable risk limits (CR101-BE1/BE2, semantics inverted by
  // CR129): `null` = no explicit override — the server enforces the
  // risk-profile preset for ALL seven, so `null` means "following your risk
  // profile", never "off". The number actually enforced arrives in
  // `resolved` below; it is never computed client-side (CR046).
  final double? sectorCapPct;
  final double? singleNameCapPct;
  final double? postLossCooldownHours;
  final int? maxOpenPositions;
  final int? maxTradesPerDay;
  final int? maxTradesPerWeek;
  final double? maxOpenRiskPct;

  // CR129: server-resolved enforced values for the seven limits above.
  // Null when the backend predates DEF193 — unknown, not "off".
  final ResolvedCaps? resolved;

  // CR129 items 2-3: the server-served Day Trader preset. Null when the
  // backend predates 568bd360 — UNKNOWN, so the Settings picker hides the
  // Day Trader entry entirely rather than approximating it client-side.
  final DayTraderPresetInfo? dayTraderPreset;
  final String learningStyle;
  final ComplianceFlags compliance;
  final String plan;
  final DateTime? trialExpiresAt;
  final int creditBalance;
  // CR039/CR047 credit state (stamped by the mandate API from `users`).
  final int creditAllowance;
  final int roomCost;
  final DateTime? creditsResetAt;
  // CR047 "The Winzip": when set and in the future, the next Room convene is in
  // cooldown — the UI can pre-empt with the countdown card. NULL/past = clear.
  final DateTime? roomCooldownUntil;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  /// True when a Winzip cooldown is currently blocking a convene.
  bool get roomCooldownActive =>
      roomCooldownUntil != null && roomCooldownUntil!.isAfter(DateTime.now());

  factory UserMandate.fromJson(Map<String, dynamic> j) {
    return UserMandate(
      userId: j['user_id'] as String,
      version: (j['version'] as num?)?.toInt() ?? 1,
      displayName: (j['display_name'] as String?) ?? 'Trader',
      locale: (j['locale'] as String?) ?? 'en',
      timezone: (j['timezone'] as String?) ?? 'UTC',
      primaryGoal: j['primary_goal'] as String? ?? 'long_term_wealth',
      horizon: j['horizon'] as String? ?? 'long',
      targetOutcome: j['target_outcome'] != null
          ? TargetOutcome.fromJson(
              (j['target_outcome'] as Map).cast<String, dynamic>(),
            )
          : null,
      path: j['path'] as String? ?? 'long_horizon',
      riskScore: (j['risk_score'] as num?)?.toInt() ?? 3,
      riskComponents: RiskComponents.fromJson(
        (j['risk_components'] as Map?)?.cast<String, dynamic>() ?? const {},
      ),
      riskQuotes:
          ((j['risk_quotes'] as List?) ?? const []).cast<String>(),
      maxDrawdownPct: (j['max_drawdown_pct'] as num?)?.toInt() ?? 30,
      sectorCapPct: (j['sector_cap_pct'] as num?)?.toDouble(),
      singleNameCapPct: (j['single_name_cap_pct'] as num?)?.toDouble(),
      postLossCooldownHours: (j['post_loss_cooldown_hours'] as num?)?.toDouble(),
      maxOpenPositions: (j['max_open_positions'] as num?)?.toInt(),
      maxTradesPerDay: (j['max_trades_per_day'] as num?)?.toInt(),
      maxTradesPerWeek: (j['max_trades_per_week'] as num?)?.toInt(),
      maxOpenRiskPct: (j['max_open_risk_pct'] as num?)?.toDouble(),
      resolved: j['resolved'] != null
          ? ResolvedCaps.fromJson((j['resolved'] as Map).cast<String, dynamic>())
          : null,
      dayTraderPreset: j['day_trader_preset'] != null
          ? DayTraderPresetInfo.fromJson(
              (j['day_trader_preset'] as Map).cast<String, dynamic>(),
            )
          : null,
      learningStyle: j['learning_style'] as String? ?? 'quick',
      compliance: ComplianceFlags.fromJson(
        (j['compliance'] as Map?)?.cast<String, dynamic>() ?? const {},
      ),
      plan: j['plan'] as String? ?? 'trial_trader',
      trialExpiresAt: j['trial_expires_at'] != null
          ? DateTime.parse(j['trial_expires_at'] as String)
          : null,
      creditBalance: (j['credit_balance'] as num?)?.toInt() ?? 75,
      creditAllowance: (j['credit_allowance'] as num?)?.toInt() ?? 0,
      roomCost: (j['room_cost'] as num?)?.toInt() ?? 0,
      creditsResetAt: j['credits_reset_at'] != null
          ? DateTime.parse(j['credits_reset_at'] as String).toLocal()
          : null,
      roomCooldownUntil: j['room_cooldown_until'] != null
          ? DateTime.parse(j['room_cooldown_until'] as String).toLocal()
          : null,
      createdAt: j['created_at'] != null
          ? DateTime.parse(j['created_at'] as String)
          : null,
      updatedAt: j['updated_at'] != null
          ? DateTime.parse(j['updated_at'] as String)
          : null,
    );
  }
}

File: backend/app/schemas/mandate.py
"""Mandate schema — the user's financial profile injected into every agent's prompt.

See docs/initial_specs/03_onboarding/mandate_schema.md for the full spec.
"""

from datetime import datetime
from enum import Enum
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class PrimaryGoal(str, Enum):
    RETIREMENT = "retirement"
    LONG_TERM_WEALTH = "long_term_wealth"
    INCOME_NOW = "income_now"
    SPECIFIC_GOAL = "specific_goal"
    LEARNING_TO_TRADE = "learning_to_trade"
    EXPLORING = "exploring"


class Horizon(str, Enum):
    SHORT = "short"
    MEDIUM = "medium"
    LONG = "long"
    VERY_LONG = "very_long"


class Path(str, Enum):
    ACTIVE = "active"
    LONG_HORIZON = "long_horizon"
    BOTH = "both"


class LearningStyle(str, Enum):
    QUICK = "quick"
    STORY = "story"
    VISUAL = "visual"
    HANDS_ON = "hands_on"


class Plan(str, Enum):
    FLOOR_PASS = "floor_pass"
    TRADER = "trader"
    FLOOR_MANAGER = "floor_manager"
    TRIAL_TRADER = "trial_trader"


# DEF179: these Mandate fields describe entitlement state that actually lives
# on `users` (see `Mandate.plan` docstring below) and is resolved server-side
# via `effective_plan_for_user()` / `balance_for()`. A client-supplied value
# for any of them — via `PATCH /v1/mandate/{user_id}` or a Brief/1-on-1
# `mandate_override` body — is dropped rather than honoured, silently: the
# server always re-stamps the real value on read (`_with_plan_state`), so a
# stored/overridden client value would be invisible on the next GET anyway.
CLIENT_UNWRITABLE_MANDATE_FIELDS = frozenset({
    "plan",
    "credit_balance",
    "credit_allowance",
    "trial_expires_at",
    "trial_started_at",
})


class TargetOutcome(BaseModel):
    amount: float
    currency: str = "USD"
    by_year: int


class RiskComponents(BaseModel):
    drawdown_response: int = Field(..., ge=1, le=5)
    regret_asymmetry: int = Field(..., ge=-1, le=1)
    concentration_tolerance: int = Field(..., ge=1, le=5)


class ResolvedCaps(BaseModel):
    """DEF193: the ENFORCED value of each preset-backed cap, server-resolved.

    `Mandate.sector_cap_pct` / `single_name_cap_pct` are `None` until a user
    explicitly overrides them, but a real, binding preset still applies —
    `GET /v1/portfolio/sector-allocation/{id}`'s `max_allowed` has shown that
    resolved number for months. Stamping the SAME resolution
    (`app.trading_math.sizing.resolved_sector_cap_pct` /
    `resolved_single_name_cap_pct` — the one place each cap is computed, per
    CR046) onto the mandate GET response closes the gap where a client had to
    choose between fabricating a number or hiding one. Same percentage-point
    units as the field each resolves; never null."""

    sector_cap_pct: float
    single_name_cap_pct: float


class Compliance(BaseModel):
    halal: bool = False
    esg_lite: bool = False
    no_tobacco_alcohol_gambling: bool = False
    no_fossil_fuels: bool = False
    long_only: bool = True
    liquid_only: bool = True
    ticker_blocklist: list[str] = Field(default_factory=list)
    ticker_allowlist: list[str] | None = None
    custom_constraints: list[str] = Field(default_factory=list)


class Mandate(BaseModel):
    """Versioned user mandate. Injected into every agent's prompt as an overlay."""

    model_config = ConfigDict(use_enum_values=True)

    user_id: UUID
    version: int = 1

    # Identity
    display_name: str
    locale: str = "en"
    timezone: str = "UTC"

    # Goal & horizon
    primary_goal: PrimaryGoal
    horizon: Horizon
    target_outcome: TargetOutcome | None = None
    path: Path

    # Risk
    risk_score: int = Field(..., ge=1, le=5)
    risk_components: RiskComponents
    risk_quotes: list[str] = Field(default_factory=list)
    max_drawdown_pct: Literal[10, 20, 30, 50, 100]

    # CR101-BE1: the two ALREADY-enforced risk caps (sector concentration, single-name
    # position size), made explicit and settable. `None` = not explicitly set — the
    # enforcement/overlay/disclosure resolvers (`sector_allocation.sector_concentration_cap`,
    # `safety_floor.single_name_cap_pct`) fall back to the preset table keyed by
    # `risk_components.concentration_tolerance` / `risk_score`, i.e. the SAME computation
    # this replaces. An existing stored mandate snapshot has no such key at all (same as
    # `None`), so it enforces IDENTICALLY post-migration. Percentage points (e.g. 40.0 =
    # 40%), matching `max_drawdown_pct`'s units. Once a PATCH sets either field it is
    # sticky — it no longer moves if risk_score/concentration_tolerance later change.
    sector_cap_pct: float | None = None
    single_name_cap_pct: float | None = None

    # CR101-BE2: four risk limits that did not exist in ANY form pre-CR101 — no
    # legacy value to migrate, so `None` simply means "off / not enforced", not
    # "fall back to a preset" (unlike sector_cap_pct/single_name_cap_pct above).
    # Each is enforced deterministically in `agents/safety_floor.py`, disclosed
    # in its own units in `agents/overlay_generator.py`, and settable via PATCH.
    # Round 2: enforced at EVERY `check_mandate_compliance` call site — the
    # direct trade-ticket path AND the Room (both the scripted path and the
    # LLM-override wrapper) — not just the ticket path round 1 shipped it on;
    # a fifth call site omitting the context now fails loudly rather than
    # silently, per the CR101-BE2 round-2 architect/auditor findings.
    # No ceiling is imposed on what a user may set (L3) — loud disclosure at
    # set-time is a mobile concern (CR101-MOBILE), out of scope here.
    #
    # Hours after a stop-out (a trade closed with a realised loss) before the
    # next BUY is allowed. Evaluated against the most recent lost trade's
    # `closed_at`.
    post_loss_cooldown_hours: float | None = None
    # Ceiling on distinct tickers concurrently held. A BUY that would open a
    # NEW position (a ticker not already held) is blocked once the count is
    # already at/above this; adding to an existing holding is unaffected.
    max_open_positions: int | None = None
    # Over-trading brake — a ceiling on trades submitted (any side) within the
    # current UTC calendar day / ISO week (Monday 00:00 UTC boundary). Fixed
    # UTC basis, not `Mandate.timezone` — see `agents/safety_floor.py` for why.
    max_trades_per_day: int | None = None
    max_trades_per_week: int | None = None
    # Sum of (position size % of portfolio) x (stop distance % below entry) /
    # 100 across open positions, in percentage points. Caps the portfolio's
    # total capital-at-risk-to-stops, not any single position. Requires a
    # stop on the proposed trade to price its own contribution; an open
    # position with no stop contributes 0 (nothing to sum).
    max_open_risk_pct: float | None = None

    # Constraints — hard rules
    compliance: Compliance = Field(default_factory=Compliance)

    # Preferences
    learning_style: LearningStyle = LearningStyle.QUICK

    # DEF129 removed `daily_briefing: DailyBriefing`. It was collected at
    # onboarding, persisted, and echoed back as a settled arrangement, while
    # nothing in the backend could schedule or deliver it. Existing mandate
    # snapshots in the `mandates` JSONB column still carry the key; Pydantic
    # ignores unknown fields, so they load unchanged and no migration is owed.

    # Plan. Not persisted on the mandate row — these live on `users` and are
    # stamped onto the response by the mandate API (CR039). Until then they
    # were schema defaults that nothing ever populated, so every client read
    # `floor_pass` / 0 credits regardless of the user's actual state.
    plan: Plan = Plan.FLOOR_PASS
    trial_expires_at: datetime | None = None
    credit_balance: int = 0
    credit_allowance: int = 0
    credits_reset_at: datetime | None = None
    room_cost: int = 0
    # CR047 "The Winzip": when set and in the future, the next Room convene is
    # in cooldown — the client shows the countdown card instead of firing a
    # doomed request. NULL / past = no cooldown pending.
    room_cooldown_until: datetime | None = None

    # DEF193: stamped by the mandate API's read path (`_with_plan_state`),
    # same as `plan`/`credit_balance` above — never persisted on the row.
    resolved: ResolvedCaps | None = None

    created_at: datetime
    updated_at: datetime
